### PR TITLE
Add FSRS-7 + Logistic Regression Ensemble model

### PR DESCRIPTION
**Pull request by HuyAiT**
## Summary

This PR adds a new experimental algorithm, **FSRS-7-LR-Ensemble**, that blends FSRS-7 and Logistic Regression predictions in logit space.

The main goal is to test whether a small linear feature-based model can complement FSRS-7 on some user histories while keeping the implementation relatively simple and the parameter count low.

Implementation details:

- Train temporary FSRS-7 and Logistic Regression models on the first 70% of the per-user training fold
- Fit same-day and non-same-day blending weights on the last 30% temporal holdout using out-of-sample predictions
- Retrain final FSRS-7 and Logistic Regression models on the full per-user training fold
- Blend final test predictions in logit space
- Regularize blending weights toward 0.4 and clamp them to `[0.1, 0.9]`

This replaces my earlier PR #332 with a cleaner branch rebased on the latest `main` and avoids fitting blending weights on predictions from models that already saw the holdout rows.

Related issue: #331

## Parameters

The logged parameter count is 71:

- 35 FSRS-7 parameters
- 34 Logistic Regression coefficients
- 2 blending weights (`w_fsrs_nsd`, `w_fsrs_sd`)

## Local validation

I ran a local validation on 500 sampled users from the 10k dataset, using seed = 2026 and explicitly including one previously problematic user (`241`). Both models were run on the same 500 users with:

```bash
python script.py --algo FSRS-7 --short --secs --equalize_test_with_non_secs
python script.py --algo FSRS-7-LR-Ensemble --short --secs --equalize_test_with_non_secs
```

Weighted by reviews on those 500 users:

| Metric | FSRS-7 | FSRS-7-LR-Ensemble | Delta |
|---|---:|---:|---:|
| LogLoss ↓ | 0.3062 | 0.3024 | -0.0038 |
| RMSE(bins) ↓ | 0.0466 | 0.0422 | -0.0044 |
| AUC ↑ | 0.7085 | 0.7158 | +0.0073 |
| MBE | 0.0034 | 0.0037 | +0.0003 |

Weighted by users on the same sample:

| Metric | FSRS-7 | FSRS-7-LR-Ensemble | Delta |
|---|---:|---:|---:|
| LogLoss ↓ | 0.3406 | 0.3359 | -0.0047 |
| RMSE(bins) ↓ | 0.0649 | 0.0604 | -0.0045 |
| AUC ↑ | 0.7048 | 0.7119 | +0.0071 |
| MBE | 0.0035 | 0.0035 | +0.0000 |

Per-user wins on the 500-user sample:

- LogLoss: 412 / 500
- RMSE(bins): 397 / 500
- AUC: 380 / 500
- absolute MBE: 342 / 500

Important caveat: this is still only a sampled local benchmark, not the full 10k-user benchmark. The ensemble also still underperforms FSRS-7 on some individual users. For example, user `241` remains worse for the ensemble, especially on AUC, so I do not want to overstate the result. The sample suggests the method may help on average, but a full benchmark is needed before drawing a strong conclusion.

## Test plan

- [x] Ran FSRS-7 and FSRS-7-LR-Ensemble through `script.py` on 500 sampled users
- [x] Included previously problematic user `241` in the sample
- [x] Confirmed the ensemble produces 71 logged parameters
- [x] Confirmed blending weights stay within `[0.1, 0.9]`
- [x] Confirmed changed Python files compile locally
- [x] Addressed Gemini review feedback about `Trainer` registration and boolean mask handling
- [ ] Full 10k-user benchmark

I am happy to revise or simplify the implementation if this approach does not fit the benchmark design, or if you would prefer a different validation protocol for ensemble/stacking models.

### CODE REVIEW

File: models/fsrs7_lr_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+    def _train_fsrs(self, df: pd.DataFrame):
+        fsrs_model = FSRS7(config=self.fsrs_config)
+        trainer = self._trainer_cls()(
+            model=fsrs_model,
+            train_set=df,
+            test_set=None,
+            batch_size=self.fsrs_config.batch_size,
+        )
+        return trainer.train()
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply models/fsrs7_lr_ensemble.py]: Addressed in commit 77d7093. I replaced the dynamic sys.modules lookup with an explicit Trainer attribute on FSRS7LREnsemble, and the model now raises a clear error if Trainer has not been registered before optimization.

File: script.py [hunk 2]
```diff
@@ -318,7 +318,7 @@ def _fit_trainable_weights(train_df: pd.DataFrame) -> Any:
         if config.device.type == "mps":
             torch.mps.empty_cache()
         return weights
-    elif config.model_name == "LogisticRegression":
+    elif config.model_name in ["LogisticRegression", "FSRS-7-LR-Ensemble"]:
         return cast(Any, model).optimize(train_df)
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply script.py:324]: Addressed in commit 77d7093. script.py now registers the benchmark Trainer on the FSRS-7-LR-Ensemble model instance before calling optimize().

File: models/fsrs7_lr_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    def predict(self, df: pd.DataFrame) -> np.ndarray:
+        fsrs_model, lr_model = self._get_sub_models()
+        from utils import Collection
+
+        fsrs_ret, _, _ = Collection(fsrs_model, self.fsrs_config).batch_predict(df)
+        lr_ret = lr_model.predict(df).cpu().numpy()
+
+        logit_fsrs = self._to_logits(np.array(fsrs_ret))
+        logit_lr = self._to_logits(np.array(lr_ret))
+        is_same_day = (df["delta_t_int"].values == 0).astype(np.float64)
```
[the remark this answers is not in the document]
Comment (HuyAiT) [reply models/fsrs7_lr_ensemble.py]: Addressed in commit 77d7093. is_same_day is now kept as a boolean array before being passed into the ensemble logit blending helper.
